feat(avm): cd_copy, rd_copy and cd_hashing

IlyasRidhuan opened the pull request:
Adds CD_COPY / RD_COPY and unconstrained calldata hashing. There's also a few changes for cd(rd)_size which is now loaded as part of the CALL and RETURN opcodes respectively.

### Calldata Handling :
* `calldata.pil` and `calldata_hashing.pil` files - calldata hashing is unconstrained
* calldata is emitted as part of public call request phase events emission

### Data Copy
* `data_copy.pil` to handles `CALLDATACOPY` and `RETURNDATACOPY` operations.
* `CD_COPY` for top level calls are looked up into the calldata column (partitioned by context id)
* Out of bounds handling is incomplete.

### Misc renames:
* Replaced `parent_calldata_offset_addr` and `parent_calldata_size_addr` with `parent_calldata_addr` and `parent_calldata_size`.
* These would eventually be used by the `X_SIZE` opcodes

IlyasRidhuan committed:
feat(avm): cd_copy

IlyasRidhuan committed:
feat: calldata and some hashing

IlyasRidhuan commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/AztecProtocol/aztec-packages/14161?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#14161** <a href="https://app.graphite.dev/github/pr/AztecProtocol/aztec-packages/14161?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/AztecProtocol/aztec-packages/14161?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#13912** <a href="https://app.graphite.dev/github/pr/AztecProtocol/aztec-packages/13912?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `master`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

sirasistant reviewed (commented):
LGTM in general, only two small points

fcarreiro commented on barretenberg/cpp/src/barretenberg/vm2/tracegen/calldata_trace.cpp:
```diff
@@ -0,0 +1,91 @@
+        }
+    }
+}
+void CalldataTraceBuilder::process_hashing(
```
nit: newline before

fcarreiro reviewed (commented):
will do a pass before it gets merged to see what this mysterious datacopy component is in simulation :)

IlyasRidhuan commented on barretenberg/cpp/pil/vm2/data_copy.pil:
```diff
@@ -0,0 +1,149 @@
+    pol commit clk;
+
+    // Constrained as inputs from execution trace
+    pol commit context_id; // Needed if we are going to read calldata from the column
```
yep good catch - it is unnecessary!

IlyasRidhuan commented on barretenberg/cpp/src/barretenberg/vm2/simulation/data_copy.cpp:
```diff
@@ -0,0 +1,59 @@
+    auto padded_calldata = context.get_calldata(cd_offset, cd_copy_size);
+
+    for (uint32_t i = 0; i < cd_copy_size; i++) {
+        memory.set(dst_addr + i, MemoryValue::from(padded_calldata[i]));
```
yep - ill add the todo, but there isnt error handling here yet

IlyasRidhuan commented on barretenberg/cpp/src/barretenberg/vm2/simulation/data_copy.cpp:
```diff
@@ -0,0 +1,59 @@
+                       const MemoryAddress dst_addr)
+{
+    auto& memory = context.get_memory();
+    auto padded_calldata = context.get_calldata(cd_offset, cd_copy_size);
```
can there be a failure at this point?

fcarreiro reviewed (commented):
Ignore what doesnt apply

IlyasRidhuan committed:
fix: add memory

Commit:
fix: better cd_hashing abstraction (#15021)

Also removes `AvmEnqueuedCallHint` in favour of bringing over the
`PublicCallRequestWithCalldata`

IlyasRidhuan committed:
fix: merge conflicts

IlyasRidhuan committed:
test data

sirasistant reviewed (approved):
LGTM on my side

fcarreiro commented on barretenberg/cpp/src/barretenberg/vm2/constraining/relations/data_copy.test.cpp:
```diff
@@ -0,0 +1,177 @@
+    }
+
+    check_relation<data_copy>(trace);
+    tracegen::LookupIntoDynamicTableGeneric<mem_read_lookup::Settings>().process(trace);
```
No need to do `::Settings`, there is a _settings class defined IIRC

fcarreiro commented on barretenberg/cpp/pil/vm2/data_copy.pil:
```diff
@@ -0,0 +1,175 @@
+    #[MEM_READ]
+    sel_mem_read { clk, read_addr, value, precomputed.zero, precomputed.zero, src_context_id }
+    in
+    memory.sel { memory.clk, memory.address, memory.value, memory.tag, memory.rw, memory.space_id};
```
nit: missing ` ` before `};`

fcarreiro commented on barretenberg/cpp/pil/vm2/data_copy.pil:
```diff
@@ -0,0 +1,175 @@
+    SEL * padding * value = 0;
+
+    #[MEM_READ]
+    sel_mem_read { clk, read_addr, value, precomputed.zero, precomputed.zero, src_context_id }
```
could you add comments in the `precomputed.zero`s? Like `/*rw=*/precomputed.zero/*(read)*/` and the other one is an FF?

fcarreiro commented on barretenberg/cpp/pil/vm2/data_copy.pil:
```diff
@@ -0,0 +1,175 @@
+    #[INCR_WRITE_ADDR]
+    (1 - precomputed.first_row) * SEL * (1 - sel_end) * (write_addr' - write_addr - 1) = 0;
+    #[MEM_WRITE]
+    sel_mem_write { clk, write_addr, value, precomputed.zero, one, dst_context_id }
```
Could you add comments to the weird cols (precomputed.zero, one). See my other comment.

BTW, you don't need `one`, you can use the selector which will always be one under these conditions `sel_mem_write`.

(check other lookups just in case)

fcarreiro commented on barretenberg/cpp/src/barretenberg/vm2/simulation/calldata_hashing.hpp:
```diff
@@ -0,0 +1,57 @@
+class CalldataHashingInterface {
+  public:
+    virtual ~CalldataHashingInterface() = default;
+    virtual FF compute_calldata_hash(std::span<const FF> calldata) = 0;
```
Nice

fcarreiro commented on barretenberg/cpp/src/barretenberg/vm2/simulation/calldata_hashing.hpp:
```diff
@@ -0,0 +1,57 @@
+    Poseidon2Interface& hasher;
+};
+
+class CalldataHashingProviderInterface {
```
☕

fcarreiro reviewed (commented):
LG in general, thanks for the changes!

IlyasRidhuan commented on barretenberg/cpp/src/barretenberg/vm2/constraining/relations/data_copy.test.cpp:
```diff
@@ -0,0 +1,177 @@
+          { C::data_copy_cd_copy_col_read, 0 },
+          { C::data_copy_cd_index, 0 } },
+    });
+}
```
oh yep, i must've removed the check_trace
